### ISSUE TITLE
Empty `belongsTo` field in cache

### ISSUE THREAD

**Issue by Jopie01** [emberjs/data#8568]
I have a case where the content of different fields depends on each other to filter out records which are not possible. The filter is used to get the right data from the backend.

For example:
1. I select a country
2. A request is send to the backend to fetch the subdivisions (because the country changed). The country is send in the payload to get only the subdivisions for that country. The response returns also the content of the subdivision because sometimes this can be prefilled so I don't have to manually select the subdivision
3. I select a subdivision from the list if needed

Everything went well, the subdivision field is updated and changed in the cache. However:

4. I selected the wrong country so I change the country
6. See 2. Besides the country the subdivision is also send.
7. In the response the subdivision is `null` because the backend cannot match the send subdivision with the country
8. The subdivision field stays filled with the 'old' subdivision

I hoped it would be possible to empty the subdivision field. When I look into the `graph` package I see that you can remove related records, but just from a `hasMany` and not from a `belongsTo`. Everything happens in the cache so I don't have any logic to cover this.

It would be nice to have the possibility build-in to empty the `belongsTo` relation.

BTW, is this sentence correct?
https://github.com/emberjs/data/blob/0c11db270b19d0ece17530e2d8b1410dde0b64f3/packages/graph/src/-private/graph/operations/remove-from-related-records.ts#L39
Or should is be something like `expected an identifier to be removed from the relationship`

**Comment by Jopie01** [comment]
Feels like a hack but when I convert the payload to JSON-API I get something like:
```javascript
data: [{
  id: 45,
  type: 'party.address',
  attributes: { },
  relationships: {
    country: {
      id: 1,
      type: 'country',
      attributes: { }
    },
    subdivision: {
      id: null,
      type: 'country.subdivision'
    }
  }
}]
```
So when the relationships are handled in:
https://github.com/emberjs/data/blob/b9685e838491db9892404799ae319de81942d75b/packages/json-api/src/-private/cache.ts#L1289-L1315

I added a small piece of code which checks if the id is `null` and if so, it adds a `replaceRelatedRecord` operation.

```ts
function setupRelationships(
  storeWrapper: CacheStoreWrapper,
  identifier: StableRecordIdentifier,
  data: JsonApiResource
) {
  // TODO @runspired iterating by definitions instead of by payload keys
  // allows relationship payloads to be ignored silently if no relationship
  // definition exists. Ensure there's a test for this and then consider
  // moving this to an assertion. This check should possibly live in the graph.
  const relationships = storeWrapper.getSchemaDefinitionService().relationshipsDefinitionFor(identifier);
  const keys = Object.keys(relationships);
  for (let i = 0; i < keys.length; i++) {
    const relationshipName = keys[i];
    const relationshipData = data.relationships![relationshipName];

    if (!relationshipData) {
      continue;
    }
    if (relationshipData.data.id === null) {
      graphFor(storeWrapper).push({
        op: 'replaceRelatedRecord',
        record: identifier,
        field: relationshipName,
        value: null
      });
      continue;
    }
    graphFor(storeWrapper).push({
      op: 'updateRelationship',
      record: identifier,
      field: relationshipName,
      value: relationshipData,
    });
  }
}
```
I know for sure you will come up with a much better and much more robust idea, but this works for me very reliably.

**Comment by runspired** [comment]
```ts
data: [{
  id: 45,
  type: 'party.address',
  attributes: { },
  relationships: {
    country: {
      id: 1,
      type: 'country',
      attributes: { }
    },
    subdivision: {
      id: null,
      type: 'country.subdivision'
    }
  }
}]
```

note this isn't the right JSON:API format and that's probably why this doesn't just work. The value for `subdivision` should just be `null`

There's a chance this is occurring because of using the wrong graph operation. If so I'd need more information, and would be happy to pair. I'd really like to get the graph package firmed up and made public soon 😅 (direct usage of it is not public yet but I'm hoping to get it there before 6.0). I would expect based on you extending the JSON:API Cache that you don't manually generate any graph operations for the "didCommit" following a save or "put" following a GET, only for "mutate". There's a chance it's the mutate operation going wrong.

**Comment by Jopie01** [comment]
> The value for `subdivision` should just be null

When I do that, the data is ignored because of
https://github.com/emberjs/data/blob/0c11db270b19d0ece17530e2d8b1410dde0b64f3/packages/json-api/src/-private/cache.ts#L1304
so no notification that the record is updated. The `subdivision` should still be in the `relations` object right?

> ... I'd need more information ...

What I really do is converting my JSONRPC payload into a JSONAPI payload and then I call `super` to let `@ember-data` do the rest. I pick it up again when I `instantiate` the record or getting the change notification.

```javascript
// extending the JSON-API cache
class RPCCache extends Cache {

  // convert the JSON response into a JSONAPI response
  convertRPC(payload, model) {
    .....

    const result = {
      data: res,
      included: includes
    }
    return result;
  }

  put(document) {
    const method = document.request.data.method.split('.');
    const model = method.slice(1, method.length - 1).join('.');
    let content = document.content

    // convert the JSON-RPC payload from the backend to the JSON-API payload
    const payload = this.convertRPC(content, model);

    document.content = payload;
    return super.put(...arguments);
  }
}

// Example of my converted payload
{
  "data": [
    {
      "type": "party.address",
      "id": 36,
      "attributes": {
        "id": 36
      },
      "relationships": {
        "subdivision": null
      }
    }
  ],
  "included": []
}
```
There are no manual changes. The usecase I described is called an `on_change`. So when a field changes it will notify the backend that it has changed. Then the backend returns an object with data as response. That data is then fed into `@ember-data` which then notifies the change and my record magically updates the fields which were in the response. This works perfectly for all usecases. Sometimes 6 fields has to be updated which happens without any problem, but when a field should be emptied (backend returns a `null` for that field) that fails.

So in short:
1. field is empty at the start
2. field is filled based on another field which filtered out the records
3. field must be emptied because that other field has changed which notified the backend who returned that the data of this field should be `null`. However this is never pushed into the cache and no notification.

**Comment by runspired** [comment]
```ts
// Example of my converted payload
{
  "data": [
    {
      "type": "party.address",
      "id": 36,
      "attributes": {
        "id": 36
      },
      "relationships": {
        "subdivision": null
      }
    }
  ],
  "included": []
}
```


this is still not valid JSON:API, I'll fix it for you maybe that's the entire issue here 😅 

```ts
// Example of my converted payload
{
  "data": [
    {
      "type": "party.address",
      "id": 36,
      "attributes": {
        "id": 36
      },
      "relationships": {
        "subdivision": { data: null } // this is the key change
      }
    }
  ],
  "included": []
}
```

**Comment by Jopie01** [comment]
>  that's the entire issue here
> "subdivision": { data: null } // this is the key change

...... don't have to say anything anymore ..... That was the culprit.
